File: src/registroestudiantes/ui/Menu.java
package registroestudiantes.ui;

import registroestudiantes.model.Estudiante;
import registroestudiantes.tda.RegistroEstudiantes;
import java.util.Scanner;

public class Menu {
    private final Scanner scanner;
    private final RegistroEstudiantes registro;

    public Menu(Scanner scanner, RegistroEstudiantes registro) {
        this.scanner = scanner;
        this.registro = registro;
    }

    public void mostrar() {
        System.out.println("\n" + "=".repeat(80));
        System.out.println("SISTEMA DE REGISTRO DE ESTUDIANTES - CURSO 3B");
        System.out.println("Capacidad: " + registro.getCapacidad() + " estudiantes");
        System.out.println("=".repeat(80));
        System.out.println("1. Registrar estudiante");
        System.out.println("2. Listar estudiantes");
        System.out.println("3. Buscar estudiante por ID");
        System.out.println("4. Modificar estudiante");
        System.out.println("5. Eliminar estudiante");
        System.out.println("6. Verificar integridad del arreglo");
        System.out.println("0. Salir");
        System.out.println("=".repeat(80));
        System.out.print("Seleccione una opcion: ");
    }

    public void registrar() {
        if (registro.estaLleno()) {
            System.out.println("Error: Capacidad maxima alcanzada.");
            return;
        }
        System.out.println("\n--- REGISTRAR ESTUDIANTE ---");
        try {
            long id = leerLong("ID (8-10 digitos): ");
            if (registro.buscarPorId(id) != null) {
                System.out.println("Error: ID ya registrado.");
                return;
            }
            String nombre = leerString("Nombre: ");
            int edad = leerInt("Edad (15-100): ", 15, 100);
            double promedio = leerDouble("Promedio (0.0-10.0): ", 0.0, 10.0);
            Estudiante nuevo = new Estudiante(id, nombre, edad, promedio);
            registro.registrar(nuevo);
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        }
    }

    public void listar() {
        registro.listarTodos();
    }

    public void buscar() {
        if (registro.estaVacio()) {
            System.out.println("El registro esta vacio.");
            return;
        }
        long id = leerLong("ID a buscar: ");
        Estudiante est = registro.buscarPorId(id);
        if (est != null) {
            System.out.println("Estudiante encontrado:");
            System.out.println(est);
        } else {
            System.out.println("No encontrado.");
        }
    }

    public void modificar() {
        if (registro.estaVacio()) {
            System.out.println("El registro esta vacio.");
            return;
        }
        long id = leerLong("ID a modificar: ");
        Estudiante est = registro.buscarPorId(id);
        if (est == null) {
            System.out.println("No encontrado.");
            return;
        }
        System.out.println("Datos actuales: " + est);
        System.out.println("\nIngrese nuevos datos (Enter para mantener):");
        String nombre = leerStringOpcional("Nuevo nombre: ", est.getNombre());
        int edad = leerIntOpcional("Nueva edad (15-100): ", est.getEdad(), 15, 100);
        double promedio = leerDoubleOpcional("Nuevo promedio (0.0-10.0): ", est.getPromedio(), 0.0, 10.0);
        registro.modificar(id, nombre, edad, promedio);
    }

    public void eliminar() {
        if (registro.estaVacio()) {
            System.out.println("El registro esta vacio.");
            return;
        }
        long id = leerLong("ID a eliminar: ");
        Estudiante est = registro.buscarPorId(id);
        if (est == null) {
            System.out.println("No encontrado.");
            return;
        }
        System.out.println("Estudiante a eliminar: " + est);
        System.out.print("Confirmar (s/N): ");
        String confirm = scanner.nextLine().trim().toLowerCase();
        if (confirm.equals("s") || confirm.equals("si")) {
            registro.eliminar(id);
        } else {
            System.out.println("Operacion cancelada.");
        }
    }

    public void verificar() {
        registro.verificarIntegridad();
    }

    private String leerString(String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String input = scanner.nextLine().trim();
            if (!input.isEmpty()) return input;
            System.out.println("El campo no puede estar vacio.");
        }
    }

    private String leerStringOpcional(String mensaje, String actual) {
        System.out.print(mensaje);
        String input = scanner.nextLine().trim();
        return input.isEmpty() ? actual : input;
    }

    private long leerLong(String mensaje) {
        while (true) {
            try {
                System.out.print(mensaje);
                long valor = Long.parseLong(scanner.nextLine().trim());
                if (valor >= 10000000L && valor <= 9999999999L) return valor;
                System.out.println("El ID debe tener entre 8 y 10 digitos.");
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un numero valido.");
            }
        }
    }

    private int leerInt(String mensaje, int min, int max) {
        while (true) {
            try {
                System.out.print(mensaje);
                int valor = Integer.parseInt(scanner.nextLine().trim());
                if (valor >= min && valor <= max) return valor;
                System.out.println("El valor debe estar entre " + min + " y " + max + ".");
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un numero entero.");
            }
        }
    }

    private int leerIntOpcional(String mensaje, int actual, int min, int max) {
        while (true) {
            System.out.print(mensaje);
            String input = scanner.nextLine().trim();
            if (input.isEmpty()) return actual;
            try {
                int valor = Integer.parseInt(input);
                if (valor >= min && valor <= max) return valor;
                System.out.println("El valor debe estar entre " + min + " y " + max + ".");
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un numero entero.");
            }
        }
    }

    private double leerDouble(String mensaje, double min, double max) {
        while (true) {
            try {
                System.out.print(mensaje);
                double valor = Double.parseDouble(scanner.nextLine().trim());
                if (valor >= min && valor <= max) return valor;
                System.out.println("El valor debe estar entre " + min + " y " + max + ".");
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un numero decimal.");
            }
        }
    }

    private double leerDoubleOpcional(String mensaje, double actual, double min, double max) {
        while (true) {
            System.out.print(mensaje);
            String input = scanner.nextLine().trim();
            if (input.isEmpty()) return actual;
            try {
                double valor = Double.parseDouble(input);
                if (valor >= min && valor <= max) return valor;
                System.out.println("El valor debe estar entre " + min + " y " + max + ".");
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un numero decimal.");
            }
        }
    }
}

File: src/registroestudiantesc++/ui/Menu.cpp
#include "Menu.hpp"
#include "../model/Estudiante.hpp"
#include <iomanip>
#include <limits>
#include <stdexcept>

Menu::Menu(std::istream& input, std::ostream& output, RegistroEstudiantes& registro)
    : input(input), output(output), registro(registro) {}

void Menu::mostrar() const {
    output << "\n" << std::string(80, '=') << std::endl;
    output << "SISTEMA DE REGISTRO DE ESTUDIANTES - CURSO 3B" << std::endl;
    output << "Capacidad: " << registro.getCapacidad() << " estudiantes" << std::endl;
    output << std::string(80, '=') << std::endl;
    output << "1. Registrar estudiante" << std::endl;
    output << "2. Listar estudiantes" << std::endl;
    output << "3. Buscar estudiante por ID" << std::endl;
    output << "4. Modificar estudiante" << std::endl;
    output << "5. Eliminar estudiante" << std::endl;
    output << "6. Verificar integridad del arreglo" << std::endl;
    output << "0. Salir" << std::endl;
    output << std::string(80, '=') << std::endl;
    output << "Seleccione una opcion: ";
}

void Menu::registrar() {
    if (registro.estaLleno()) {
        output << "Error: Capacidad maxima alcanzada." << std::endl;
        return;
    }
    output << "\n--- REGISTRAR ESTUDIANTE ---" << std::endl;
    try {
        long long id = leerLong("ID (8-10 digitos): ");
        if (registro.buscarPorId(id) != nullptr) {
            output << "Error: ID ya registrado." << std::endl;
            return;
        }
        std::string nombre = leerString("Nombre: ");
        int edad = leerInt("Edad (15-100): ", 15, 100);
        double promedio = leerDouble("Promedio (0.0-10.0): ", 0.0, 10.0);
        Estudiante* nuevo = new Estudiante(id, nombre, edad, promedio);
        registro.registrar(nuevo);
    } catch (const std::invalid_argument& e) {
        output << "Error: " << e.what() << std::endl;
    }
}

void Menu::listar() const {
    registro.listarTodos();
}

void Menu::buscar() const {
    if (registro.estaVacio()) {
        output << "El registro esta vacio." << std::endl;
        return;
    }
    long long id = leerLong("ID a buscar: ");
    Estudiante* est = registro.buscarPorId(id);
    if (est != nullptr) {
        output << "Estudiante encontrado:" << std::endl;
        output << *est << std::endl;
    } else {
        output << "No encontrado." << std::endl;
    }
}

void Menu::modificar() {
    if (registro.estaVacio()) {
        output << "El registro esta vacio." << std::endl;
        return;
    }
    long long id = leerLong("ID a modificar: ");
    Estudiante* est = registro.buscarPorId(id);
    if (est == nullptr) {
        output << "No encontrado." << std::endl;
        return;
    }
    output << "Datos actuales: " << *est << std::endl;
    output << "\nIngrese nuevos datos (Enter para mantener):" << std::endl;
    std::string nombre = leerStringOpcional("Nuevo nombre: ", est->getNombre());
    int edad = leerIntOpcional("Nueva edad (15-100): ", est->getEdad(), 15, 100);
    double promedio = leerDoubleOpcional("Nuevo promedio (0.0-10.0): ", est->getPromedio(), 0.0, 10.0);
    registro.modificar(id, nombre, edad, promedio);
}

void Menu::eliminar() {
    if (registro.estaVacio()) {
        output << "El registro esta vacio." << std::endl;
        return;
    }
    long long id = leerLong("ID a eliminar: ");
    Estudiante* est = registro.buscarPorId(id);
    if (est == nullptr) {
        output << "No encontrado." << std::endl;
        return;
    }
    output << "Estudiante a eliminar: " << *est << std::endl;
    output << "Confirmar (s/N): ";
    std::string confirm;
    std::getline(input, confirm);
    if (confirm == "s" || confirm == "si") {
        registro.eliminar(id);
    } else {
        output << "Operacion cancelada." << std::endl;
    }
}

void Menu::verificar() const {
    registro.verificarIntegridad();
}

std::string Menu::leerString(const std::string& mensaje) {
    while (true) {
        output << mensaje;
        std::string input;
        std::getline(this->input, input);
        if (!input.empty()) return input;
        output << "El campo no puede estar vacio." << std::endl;
    }
}

std::string Menu::leerStringOpcional(const std::string& mensaje, const std::string& actual) {
    output << mensaje;
    std::string input;
    std::getline(this->input, input);
    return input.empty() ? actual : input;
}

long long Menu::leerLong(const std::string& mensaje) {
    while (true) {
        try {
            output << mensaje;
            std::string input;
            std::getline(this->input, input);
            if (input.empty()) {
                output << "El ID no puede estar vacio." << std::endl;
                continue;
            }
            long long valor = std::stoll(input);
            if (valor >= 10000000LL && valor <= 9999999999LL) return valor;
            output << "El ID debe tener entre 8 y 10 digitos." << std::endl;
        } catch (const std::exception&) {
            output << "Debe ingresar un numero valido." << std::endl;
        }
    }
}

int Menu::leerInt(const std::string& mensaje, int min, int max) {
    while (true) {
        try {
            output << mensaje;
            std::string input;
            std::getline(this->input, input);
            if (input.empty()) {
                output << "El campo no puede estar vacio." << std::endl;
                continue;
            }
            int valor = std::stoi(input);
            if (valor >= min && valor <= max) return valor;
            output << "El valor debe estar entre " << min << " y " << max << "." << std::endl;
        } catch (const std::exception&) {
            output << "Debe ingresar un numero entero." << std::endl;
        }
    }
}

int Menu::leerIntOpcional(const std::string& mensaje, int actual, int min, int max) {
    while (true) {
        output << mensaje;
        std::string input;
        std::getline(this->input, input);
        if (input.empty()) return actual;
        try {
            int valor = std::stoi(input);
            if (valor >= min && valor <= max) return valor;
            output << "El valor debe estar entre " << min << " y " << max << "." << std::endl;
        } catch (const std::exception&) {
            output << "Debe ingresar un numero entero." << std::endl;
        }
    }
}

double Menu::leerDouble(const std::string& mensaje, double min, double max) {
    while (true) {
        try {
            output << mensaje;
            std::string input;
            std::getline(this->input, input);
            if (input.empty()) {
                output << "El campo no puede estar vacio." << std::endl;
                continue;
            }
            double valor = std::stod(input);
            if (valor >= min && valor <= max) return valor;
            output << "El valor debe estar entre " << min << " y " << max << "." << std::endl;
        } catch (const std::exception&) {
            output << "Debe ingresar un numero decimal." << std::endl;
        }
    }
}

double Menu::leerDoubleOpcional(const std::string& mensaje, double actual, double min, double max) {
    while (true) {
        output << mensaje;
        std::string input;
        std::getline(this->input, input);
        if (input.empty()) return actual;
        try {
            double valor = std::stod(input);
            if (valor >= min && valor <= max) return valor;
            output << "El valor debe estar entre " << min << " y " << max << "." << std::endl;
        } catch (const std::exception&) {
            output << "Debe ingresar un numero decimal." << std::endl;
        }
    }
}